### PR TITLE
inconsistencies part 2/2 : cyclic inconsistencies (#671)

### PR DESCRIPTION
**Pull request by glerzing**
Please double-check that there is no mistake in the function _count_cycles_and_comparison_trios.

I actually prefer to avoid also implementing an API endpoint that returns a single cycle, but of possibly longer length, because:
- it requires time and effort, and complicates things
- It would be complicated to display on the frontend
- It doesn't provide overview on how many cycles are left (we can't computationally count the cycles of any length, this would be NP-complex)

### CODE REVIEW

**Comment by lfaucon** [comment]
> I actually prefer to avoid also implementing an API endpoint that returns a single cycle, but of possibly longer length, because:

> it requires time and effort, and complicates things
> It would be complicated to display on the frontend
> It doesn't provide overview on how many cycles are left (we can't computationally count the cycles of any length, this would be NP-complex)

Agreed & Fine by me. The choice to not do something that would take time can always be changed in the future in any case. 

Thanks for this initial version that brings 99% of the value of the feature for cycle detection :)

File: backend/tournesol/views/inconsistencies.py
Comment (lfaucon) [inline backend/tournesol/views/inconsistencies.py:166]: Why do you need the `copy` operation here? Isn't it very slow to do this copy for every entity?

File: backend/tournesol/views/inconsistencies.py
Comment (lfaucon) [inline backend/tournesol/views/inconsistencies.py:132]: Using `defaultdict` would simplify the code as you would not need to initialize all dictionaries.

File: backend/tournesol/views/inconsistencies.py
Comment (lfaucon) [inline backend/tournesol/views/inconsistencies.py:168]: The code would feel more readable to me with something like this:
```
                for entity_3 in self.outgoing_connections[entity_2]:
                    if entity_1 in self.outgoing_connections[entity_3]:
```
I don't think it reduces performance, does it?
Maybe we don't need to keep track of incoming edges with that change.

File: backend/tournesol/views/inconsistencies.py
Comment (lfaucon) [inline backend/tournesol/views/inconsistencies.py:179]: Mmmm, I see. Having these operations here makes it much harder for me to keep track of what's hapening during the multiple loops.

I would find something that does not modify the list of connections but keeps track of a `visited` object simpler to reason about.

What do you think?

**Review by lfaucon — COMMENTED** [review]
This looks quite good to me. I made a few comments but they are minor. I would prefer a second pair of eyes on this feature before merging it. Maybe @farnots or @amatissart would be good candidates for reviewing.

File: backend/tournesol/views/inconsistencies.py
Comment (glerzing) [reply backend/tournesol/views/inconsistencies.py:166]: It was a problem to loop over a set, and modify it inside the loop. I think it was this error:
RuntimeError: Set changed size during iteration

File: backend/tournesol/views/inconsistencies.py
Comment (glerzing) [reply backend/tournesol/views/inconsistencies.py:168]: Actually, the & operator between sets is very efficient, especially when the sets have few elements in common. I made some experiments on python.

Comment (glerzing) [reply backend/tournesol/views/inconsistencies.py:168]: Honestly, the sets in incoming_connections and outgoing_connections must be pretty small in practice, so I'm not sure that the gain is significant, but I prefer to keep it like this.

File: backend/tournesol/views/inconsistencies.py
Comment (glerzing) [reply backend/tournesol/views/inconsistencies.py:132]: agreed

File: backend/tournesol/views/inconsistencies.py
Comment (glerzing) [reply backend/tournesol/views/inconsistencies.py:166]: outgoing_connections[entity_1] should usually contain less than a dozen of entities, so it will be pretty small copies.

File: backend/tournesol/views/inconsistencies.py
Comment (glerzing) [inline backend/tournesol/views/inconsistencies.py:150]: I shouldn't have put an "else: continue" there. all_connections is used for comparison trios, for which the score doesn't matter.

File: backend/tournesol/views/inconsistencies.py
Comment (glerzing) [reply backend/tournesol/views/inconsistencies.py:179]: There were 2 ways I think to do this:
1 - using simple for loops and if (the suggestion of the previous comment), which is better for readability 
2 - using the "&" operator on sets, which is better for performances

I don't know if I can keep using the 2nd method without progressively "removing the edges" where all the cycles were counted, even if it's not great for readability. So I believe it is a choice between these 2 techniques.

Comment (amatissart) [reply backend/tournesol/views/inconsistencies.py:179]: I agree that using immutable connections here would help to reason about the algorithm, especially for future readers and maintainers.

If "removing the edges" is only used to avoid counting the same cycle multiple times, then an alternative solution would be to consider only the cycles where `entity_1` has the smallest uid among the 3 (i.e adding conditions `entity_1 < entity_2` and `entity_1 < entity_3` when iterating on the connections).

(This also applies to "comparison trios").

File: backend/tournesol/serializers/inconsistencies.py
Comment (amatissart) [inline backend/tournesol/serializers/inconsistencies.py:9]: For consistency with existing API responses, I would prefer to use `criteria` here.   
(We are not very consistent across the codebase :grimacing:, but in all serializers at least `criteria` seems to be used as an invariable noun).

File: backend/tournesol/views/inconsistencies.py
Comment (lfaucon) [reply backend/tournesol/views/inconsistencies.py:166]: Ok

File: backend/tournesol/views/inconsistencies.py
Comment (lfaucon) [reply backend/tournesol/views/inconsistencies.py:168]: Agreed! Thanks

File: backend/tournesol/serializers/inconsistencies.py
Comment (glerzing) [reply backend/tournesol/serializers/inconsistencies.py:9]: I don't think this is the right spelling in theory, but ok, it is simpler to write it as an invariable noun everywhere

File: backend/tournesol/views/inconsistencies.py
Comment (glerzing) [reply backend/tournesol/views/inconsistencies.py]: For comparisons trios, I applied a different trick. It turns out that, without removing the edges, each cycle is counted 3 times, and each comparisons trio is counted 6 times. For the comparisons trios I only need a count, so I can just divide it by 6, which is more efficient than incrementing the count only when uid1 < uid2 < uid3.

**Review by amatissart — APPROVED** [review]

**Comment by GresilleSiffle** [comment]
I just updated the OpenAPI schema in `main`. I'm going to update this branch to fix the conflict.

**Comment by GresilleSiffle** [comment]
Should we merge @amatissart @glerzing ?

**Comment by glerzing** [comment]
> Should we merge @amatissart @glerzing ?
ok

**Pull request merged**
